cpp/include/cudf/table/experimental/row_operators.cuh

```diff
@@ -0,0 +1,426 @@
+   * @param depth Optional, device array the same length as a row that contains starting depths of
+   * columns if they're nested, and 0 otherwise.
+   * @param column_order Optional, device array the same length as a row that indicates the desired
+   * ascending/descending order of each column in a row. If `nullopt`, it is assumed all columns are
+   * sorted in ascending order.
+   * @param null_precedence Optional, device array the same length as a row and indicates how null
+   * values compare to all other for every column. If `nullopt`, then null precedence would be
+   * `null_order::BEFORE` for all columns.
+   */
+  device_row_comparator(Nullate has_nulls,
+                        table_device_view lhs,
+                        table_device_view rhs,
+                        std::optional<device_span<int const>> depth                  = std::nullopt,
+                        std::optional<device_span<order const>> column_order         = std::nullopt,
+                        std::optional<device_span<null_order const>> null_precedence = std::nullopt)
+    : _lhs{lhs},
+      _rhs{rhs},
+      _nulls{has_nulls},
+      _depth{depth},
+      _column_order{column_order},
+      _null_precedence{null_precedence}
+  {
+  }
+
+ public:
+  /**
+   * @brief Checks whether the row at `lhs_index` in the `lhs` table compares
+   * lexicographically less than the row at `rhs_index` in the `rhs` table.
+   *
+   * @param lhs_index The index of row in the `lhs` table to examine
+   * @param rhs_index The index of the row in the `rhs` table to examine
+   * @return `true` if row from the `lhs` table compares less than row in the `rhs` table
+   */
+  __device__ bool operator()(size_type lhs_index, size_type rhs_index) const noexcept
+  {
+    int last_null_depth = std::numeric_limits<int>::max();
+    for (size_type i = 0; i < _lhs.num_columns(); ++i) {
+      int depth = _depth.has_value() ? (*_depth)[i] : 0;
+      if (depth > last_null_depth) { continue; }
+
+      bool ascending = _column_order.has_value() ? (*_column_order)[i] == order::ASCENDING : true;
+
+      null_order null_precedence =
+        _null_precedence.has_value() ? (*_null_precedence)[i] : null_order::BEFORE;
+
+      auto comparator =
+        element_comparator{_nulls, _lhs.column(i), _rhs.column(i), null_precedence, depth};
+
+      weak_ordering state;
+      cuda::std::tie(state, last_null_depth) =
+        cudf::type_dispatcher(_lhs.column(i).type(), comparator, lhs_index, rhs_index);
+
+      if (state == weak_ordering::EQUIVALENT) { continue; }
+
+      return state == (ascending ? weak_ordering::LESS : weak_ordering::GREATER);
+    }
+    return false;
+  }
+
+ private:
+  table_device_view _lhs;
+  table_device_view _rhs;
+  Nullate _nulls{};
+  std::optional<device_span<int const>> _depth;
+  std::optional<device_span<order const>> _column_order;
+  std::optional<device_span<null_order const>> _null_precedence;
```

Comment:
Rather than actually storing the optionals, why not make the class members the value types of the optionals, unpack the optionals in the constructor and initialize the class members based on whether the optional is set or not? Then we don't need to check them inside the comparator.

Reply:
Because these are arrays. I'm trying to avoid making dynamic device arrays filled with default values when the optional passed to the ctor is a nullopt.

Reply:
Oops I completely forgot that you were indexing into arrays rather than using a scalar value for all columns. Never mind, good call.